d3a_api_client/rest_device.py

```diff
@@ -0,0 +1,164 @@
+    return decorator
+
+
+class RestDeviceClient(APIClientInterface):
+
+    def __init__(self, simulation_id, device_id, domain_name,
+                 websockets_domain_name, is_ssl, autoregister=False):
+        self.simulation_id = simulation_id
+        self.device_id = device_id
+        self.domain_name = domain_name
+        self.retrieve_key_from_server()
+
+        if is_ssl:
+            self.pem_cert = ssl.get_server_certificate(domain_name)
+        self.dispatcher = WebsocketMessageDispatcher(self)
+        self.websocket_thread = WebsocketThread(simulation_id, device_id, self.jwt_token,
+                                                websockets_domain_name, self.dispatcher)
+        self.websocket_thread.start()
+        self.callback_thread = ThreadPoolExecutor(max_workers=5)
+        self.registered = False
+        if autoregister:
+            print(f"Trying to register")
+            self.register()
+
+    def retrieve_key_from_server(self):
+        resp = requests.post(
+            f"{self.domain_name}/api-token-auth/",
+            data=json.dumps({"username": os.environ["API_CLIENT_USERNAME"],
+                             "password": os.environ["API_CLIENT_PASSWORD"]}),
+            headers={"Content-Type": "application/json"})
+        if resp.status_code != 200:
+            logging.error(f"Request for token authentication failed with status code {resp.status_code}."
+                          f"Response body: {resp.text}")
+            return
+        self.jwt_token = json.loads(resp.text)["token"]
+
+    @property
+    def _url_prefix(self):
+        return f'{self.domain_name}/external-connection/api/{self.simulation_id}/{self.device_id}/'
+
+    def post_request(self, endpoint_suffix, data):
+        resp = requests.post(
+            f"{self._url_prefix}/{endpoint_suffix}/",
+            data=json.dumps(data),
+            headers={"Content-Type": "application/json",
+                     "Authorization": f"JWT {self.jwt_token}"})
+        if resp.status_code != 200:
+            logging.error(f"Request {endpoint_suffix} failed with status code {resp.status_code}."
+                          f"Response body: {resp.text}")
+            return False
+        return True
+
+    def get_request(self, endpoint_suffix):
+        resp = requests.post(
+            f"{self._url_prefix}/{endpoint_suffix}/",
+            headers={"Content-Type": "application/json",
+                     "Authorization": f"JWT {self.jwt_token}"})
+        if resp.status_code != 200:
+            logging.error(f"Request {endpoint_suffix} failed with status code {resp.status_code}."
+                          f"Response body: {resp.text}")
+            return False
+        return True
+
+    @logging_decorator('register')
+    def register(self, is_blocking=True):
+        if self.post_request('register', {}):
+            return_value = self.dispatcher.wait_for_command_response('register')
+            self.registered = True
```

Comment:
Cant the user can also be restricted to register? In what case self.registered should not be set to True, right?

Reply:
Indeed, I'll be checking the output of register command to fill this value, thanks